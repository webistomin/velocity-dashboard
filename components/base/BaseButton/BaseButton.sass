+b('base-button')
  width: 100%
  font: var(--text-button)
  border-radius: 4px
  transition-duration: 0.2s
  transition-timing-function: ease-out

  +m('theme')
    +v('default')
      background-color: rgb(var(--color-primary))
      color: rgb(var(--color-white))
      transition-property: opacity, background-color

      &:hover,
      &:focus
        opacity: 0.9

      &:active
        opacity: 0.8

    +v('light')
      background-color: rgb(var(--color-secondary))
      color: rgb(var(--color-primary))
      transition-property: color, background-color, opacity

      &:hover,
      &:focus
        background-color: rgb(var(--color-primary))
        color: rgb(var(--color-white))

      &:active
        opacity: 0.8

    +v('success')
      background-color: rgb(var(--color-green))
      color: rgb(var(--color-white))
      transition-property: opacity, background-color

      &:hover,
      &:focus
        opacity: 0.9

      &:active
        opacity: 0.8

    +v('gray')
      background-color: rgb(var(--color-gray-5))
      color: rgb(var(--color-gray-3))
      transition-property: background-color, color

      &:hover
        color: rgb(var(--color-white))
        background-color: rgb(var(--color-gray-4))

  +e('wrapper')
    box-sizing: border-box
    width: 100%
    height: 100%
    padding: rem(9.5px) rem(18px)
    display: flex
    align-items: center
    justify-content: center

    +media($screen-sm)
      padding-top: rem(14.5px)
      padding-bottom: rem(14.5px)

  +e('icon')
    margin-right: rem(8px)
