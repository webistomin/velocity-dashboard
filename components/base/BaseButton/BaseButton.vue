<template lang="pug">
  button(
    :type='type'
    :disabled='isDisabled'
    :class='getClassnames'
    @click='onClick'
  ).base-button.btn
    span.base-button__wrapper
      SvgIcon(
        v-if='icon'
        name='icon'
        :width='16'
        :height='16'
      ).base-button__icon
      slot.base-button__content
</template>

<script lang="ts">
  import Vue from 'vue';
  import Component from 'vue-class-component';

  const BaseButtonProps = Vue.extend({
    props: {
      type: {
        type: String,
        default: 'button',
      },
      isDisabled: {
        type: Boolean,
        default: false,
      },
      isLoading: {
        type: Boolean,
        default: false,
      },
      icon: {
        type: String,
        default: '',
      },
      theme: {
        type: String,
        default: 'default',
      },
    },
  });

  @Component({
    name: 'BaseButton',
  })
  export default class BaseButton extends BaseButtonProps {
    public onClick(): void {
      this.$emit('click');
    }

    public get getClassnames(): string[] {
      const result = [`base-button_theme_${this.theme}`];

      if (this.isDisabled) {
        result.push('base-button_disabled');
      }

      if (this.isLoading) {
        result.push('base-button_loading');
      }

      return result;
    }
  }
</script>

<style lang="sass">
  @import "BaseButton"
</style>
