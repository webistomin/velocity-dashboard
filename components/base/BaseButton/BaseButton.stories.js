import { text, boolean } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import BaseButton from './BaseButton.tsx';

export default {
  title: 'BaseButton',
  parameters: {
    options: {
      showPanel: true,
      selectedPanel: 'storybook/knobs/panel',
    },
  },
};

export const Default = () => ({
  components: { BaseButton },
  props: {
    type: {
      type: String,
      default: text('type', 'button'),
    },
    isDisabled: {
      type: Boolean,
      default: boolean('isDisabled', false),
    },
    isLoading: {
      type: Boolean,
      default: boolean('isLoading', false),
    },
    icon: {
      type: String,
      default: text('icon', ''),
    },
    theme: {
      type: String,
      default: text('theme', 'default'),
    },
  },
  methods: {
    onClick: action('emit click'),
  },
  template: `<base-button
    :type='type'
    :isDisabled='isDisabled'
    :isLoading='isLoading'
    :icon='icon'
    :theme='theme'
    @click='onClick'
  >Button</base-button>`,
});
