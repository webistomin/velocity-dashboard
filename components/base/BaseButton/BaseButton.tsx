import { Prop, Component } from 'nuxt-property-decorator';
import { VNode } from 'vue';
import { VueComponent } from 'types/vue-components';
import { ButtonTypes } from 'types/common/button-types';

import './BaseButton.sass';

export type BaseButtonThemes = 'default' | 'light' | 'success' | 'gray';

export interface IBaseButtonProps {
  type?: ButtonTypes;
  isDisabled?: boolean;
  isLoading?: boolean;
  icon?: string;
  theme?: BaseButtonThemes;
  onClick: () => void;
}

@Component({
  name: 'BaseButton',
})
export default class BaseButton extends VueComponent<IBaseButtonProps> {
  @Prop({ default: 'button' })
  private readonly type!: IBaseButtonProps['type'];

  @Prop({ default: false })
  private readonly isDisabled!: IBaseButtonProps['isDisabled'];

  @Prop({ default: false })
  private readonly isLoading!: IBaseButtonProps['isLoading'];

  @Prop({ default: '' })
  private readonly icon!: IBaseButtonProps['icon'];

  @Prop({ default: 'default' })
  private readonly theme!: IBaseButtonProps['theme'];

  public onClick(): void {
    this.$emit('click');
  }

  public get getClassnames(): string[] {
    const result = [`base-button_theme_${this.theme}`];

    if (this.isDisabled) {
      result.push('base-button_disabled');
    }

    if (this.isLoading) {
      result.push('base-button_loading');
    }

    return result;
  }

  public render(): VNode {
    return (
      <button
        type={this.type}
        class={`base-button btn ${this.getClassnames.join(' ')}`}
        disabled={this.isDisabled}
        onClick={this.onClick}>
        <span class='base-button__wrapper'>
          {this.icon ? <svg-icon name={this.icon} width={16} height={16} /> : null}
          <span class='base-button__content'>{this.$slots.default}</span>
        </span>
      </button>
    );
  }
}
