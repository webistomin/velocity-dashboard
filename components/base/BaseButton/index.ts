import BaseButton from './BaseButton.vue';

export default BaseButton;
