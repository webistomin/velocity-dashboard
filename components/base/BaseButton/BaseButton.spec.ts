import { mount } from '@vue/test-utils';
import BaseButton from './BaseButton';

describe('BaseButton', () => {
  const buttonSelector = '.base-button';
  const buttonIconSelector = '.base-button__wrapper svg-icon-stub';
  const buttonContentSelector = '.base-button .base-button__content';

  const factory = (propsData = {}, options = {}) => {
    return mount(BaseButton, {
      propsData: {
        ...propsData,
      },
      ...options,
      stubs: ['svg-icon'],
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory().html();
    expect(wrapper).toMatchSnapshot();
  });

  it('Emit click event', () => {
    const wrapper = factory();
    wrapper.vm.onClick();
    expect(wrapper.emitted().click);
  });

  it('Icon is hidden if icon prop is not provided', () => {
    const wrapper = factory();
    const icon = wrapper.find(buttonIconSelector);
    expect(icon.exists()).toBe(false);
  });

  it('Icon is visible if icon prop is provided', () => {
    const wrapper = factory({
      icon: 'icon-star',
    });
    const icon = wrapper.find(buttonIconSelector);
    expect(icon.exists()).toBe(true);
  });

  it('Set disabled class is disabled prop is provided', () => {
    const expectedResult = ['base-button_disabled'];
    const wrapper = factory({
      isDisabled: true,
    });
    const button = wrapper.find(buttonSelector);
    const classes = button.classes();
    expect(classes).toEqual(expect.arrayContaining(expectedResult));
  });

  it('Set loading class is loading prop is provided', () => {
    const expectedResult = ['base-button_loading'];
    const wrapper = factory({
      isLoading: true,
    });
    const button = wrapper.find(buttonSelector);
    const classes = button.classes();
    expect(classes).toEqual(expect.arrayContaining(expectedResult));
  });

  it('Set badge text correctly', () => {
    const expected = '123';
    const wrapper = factory(
      {},
      {
        slots: {
          default: [expected],
        },
      }
    );

    const content = wrapper.find(buttonContentSelector);
    expect(content.text()).toEqual(expected);
  });
});
