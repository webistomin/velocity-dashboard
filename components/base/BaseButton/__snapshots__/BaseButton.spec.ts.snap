// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseButton Matches snapshot 1`] = `"<button type=\\"button\\" class=\\"base-button btn base-button_theme_default\\"><span class=\\"base-button__wrapper\\"><span class=\\"base-button__content\\"></span></span></button>"`;
